### PR TITLE
Add agent role and project scoping to Dark Factory memory calls

### LINKED ISSUE #148: Add agent role and project scoping to Dark Factory memory calls
Parent epic: omniscient/dark-factory#140

## Goal
Prepare the memory system for a multi-Hermes-agent factory by preventing accidental cross-project or cross-role contamination.

## Stable project scope

```text
project=markethawk
```

## Initial role IDs

```text
factory-director
intake-triage-agent
refinement-agent
planning-agent
implementation-agent
validation-agent
code-review-agent
security-agent
deconflict-agent
ci-rescue-agent
merge-gate-agent
cost-telemetry-agent
human-liaison-agent
```

## Acceptance criteria
- [ ] Every memory read includes `project=markethawk`.
- [ ] Every memory write includes `project=markethawk`.
- [ ] Every memory read/write includes a stable `agentId`.
- [ ] Cross-agent reads are explicit rather than accidental.
- [ ] Validation/security agents do not receive hidden implementation reasoning unless explicitly allowed.
- [ ] Tests cover role/project scoping behavior.

---

— Hermes Agent / Factory Planning
Model: gpt-5.5

### PR DESCRIPTION
**Pull request by omniscient**
Closes omniscient/dark-factory#148

## Summary
Automated implementation for issue omniscient/dark-factory#148.

## Preview
⚠️ _Preview environment failed to start (preview failed: docker compose up failed (migrate/boot — see preview_build.log tail below)) — endpoint validation was skipped; see the failure comment on the issue._

## Commands
```bash
# Iterate after feedback
docker compose --profile factory run --rm dark-factory "Continue issue omniscient/dark-factory#148"

# Tear down preview when done
docker compose --profile factory run --rm dark-factory "Close issue omniscient/dark-factory#148"
```

---
*Generated by MarketHawk Dark Factory*

### CODE REVIEW

File: dark-factory/scripts/gate_lib.sh [hunk 1]
```diff
@@ -15,8 +19,9 @@ route_memory_file() {
 }
 
 write_memory_entry() {
-  # Usage: write_memory_entry TARGET PATH_PREFIX VIOLATION_TEXT SOURCE ISSUE_NUM
+  # Usage: write_memory_entry TARGET PATH_PREFIX VIOLATION_TEXT SOURCE ISSUE_NUM [AGENT_ROLE]
   local TARGET="$1" PATH_PREFIX="$2" TEXT="$3" SOURCE="$4" ISSUE="$5"
+  local ROLE="${6:-${AGENT_ID:-unknown}}"
 
   # Dedup: skip if core sentence already present
   if grep -qF "$TEXT" "$TARGET" 2>/dev/null; then
```
Comment (omniscient) [inline dark-factory/scripts/gate_lib.sh]: **correctness** · **medium** — The dedup `grep -qF "$TEXT"` precedes tagging, so identical TEXT submitted later by a different project/agentId is dropped and the first writer's scope is the only one persisted, undermining cross-project separation. Consider including project/agentId in the dedup key.

File: dark-factory/tests/test_load_memory.sh [hunk 2]
```diff
@@ -15,11 +15,23 @@ assert() {
...
+REPO_ROOT=$(git rev-parse --show-toplevel)
+source "${REPO_ROOT}/dark-factory/scripts/agent_roles.sh"
 
 load_memory() {
   local MEMFILE="$1"
   [ -f "$MEMFILE" ] || return
   while IFS= read -r line; do
+    # Project filter: skip entries tagged for a different project.
+    # Entries without any project: tag are always included (legacy backward compat).
+    if echo "$line" | grep -q 'project:'; then
```
Comment (omniscient) [inline dark-factory/tests/test_load_memory.sh]: **edge-case** · **medium** — The presence check `grep -q 'project:'` scans the entire line including the human-readable TEXT; an avoidance note that mentions `project:` in prose is treated as a tagged entry and filtered. Restrict detection to the `<!-- ... -->` metadata only.

**Review by omniscient — COMMENTED** [review]
## 🏭 Dark Factory Code Review

**0** blocking · **5** advisory finding(s).

File: dark-factory/tests/test_load_memory.sh [hunk 3]
```diff
@@ -15,11 +15,23 @@ assert() {
...
+      if ! echo "$line" | grep -q "project:${MEMORY_PROJECT}"; then
```
Comment (omniscient) [inline dark-factory/tests/test_load_memory.sh]: **correctness** · **medium** — `grep -q "project:${MEMORY_PROJECT}"` is an unanchored substring match; entries tagged `project:markethawk-staging` (or any `markethawk*` prefix) wrongly pass the filter. Anchor on the tag boundary, e.g. `grep -q "project:${MEMORY_PROJECT}\b"` or match within the metadata block.

File: dark-factory/scripts/gate_lib.sh [hunk 4]
```diff
@@ -42,7 +47,7 @@ write_memory_entry() {
   fi
 
   EXPIRES=$(date -d '+6 months' +%Y-%m-%d 2>/dev/null || date -v+6m +%Y-%m-%d)
-  ENTRY="- [AVOID] $TEXT <!-- issue:#$ISSUE date:$(date +%Y-%m-%d) expires:$EXPIRES source:$SOURCE path:$PATH_PREFIX -->"
+  ENTRY="- [AVOID] $TEXT <!-- project:${MEMORY_PROJECT} agentId:${ROLE} issue:#$ISSUE date:$(date +%Y-%m-%d) expires:$EXPIRES source:$SOURCE path:$PATH_PREFIX -->"
```
Comment (omniscient) [inline dark-factory/scripts/gate_lib.sh]: **maintainability** · **low** — `${MEMORY_PROJECT}` and `${ROLE}` are inlined into the comment metadata unvalidated; a role/project containing `-->` or a newline would break the metadata structure. Constrain or sanitize these values.

File: dark-factory/tests/test_load_memory.sh [hunk 5]
```diff
@@ -15,11 +15,23 @@ assert() {
 }
 
 AFFECTED="backend/app/routers/scanner.py"
+MEMORY_PROJECT="markethawk"
```
Comment (omniscient) [inline dark-factory/tests/test_load_memory.sh]: **maintainability** · **low** — `MEMORY_PROJECT="markethawk"` on line 18 is immediately overwritten by sourcing `agent_roles.sh` on line 21 (and re-set at line 111); remove the redundant manual assignments to avoid confusion.

**Review by omniscient — COMMENTED** [review]
## Advisory Findings Addressed

Automatically addressed **5** advisory finding(s):

- [MEDIUM] test_load_memory.sh:29 (edge-case): Replaced bare `grep -q 'project:'` with metadata-only extraction via `grep -o '<!--[^>]*-->'`; prose containing `project:` no longer triggers the filter.
- [MEDIUM] test_load_memory.sh:30 (correctness): Changed project match to `grep -qE "project:${MEMORY_PROJECT}([[:space:]]|-->)"` to anchor the boundary; `markethawk-staging` no longer matches `markethawk`.
- [LOW] test_load_memory.sh:18 (maintainability): Removed redundant `MEMORY_PROJECT="markethawk"` assignments (line 18 and R3 block) that were immediately overwritten by `agent_roles.sh`.
- [MEDIUM] gate_lib.sh:27 (correctness): Replaced single-text dedup with compound check requiring both `project:${MEMORY_PROJECT}` and `agentId:${ROLE}`; identical TEXT from a different project/agent is no longer silently dropped.
- [LOW] gate_lib.sh:50 (maintainability): Added `SAFE_PROJECT`/`SAFE_ROLE` locals that strip newlines and `-->` sequences before interpolation into the HTML comment metadata.

---
*Posted by MarketHawk Dark Factory*
